parallel-exec/lib/lib.go

```diff
@@ -0,0 +1,418 @@
+
+	defaultFastFail = false
+)
+
+var (
+	defaultMaxConcurrentCmds = runtime.NumCPU()
+	defaultEventHandler      = func(event Event) {
+		data, err := json.Marshal(event)
+		if err != nil {
+			log.Print(event.Type())
+		}
+		log.Print(string(data))
+	}
+	defaultTimer = time.Now
+
+	errCmdFailed   = errors.New("command failed")
+	errInterrupted = errors.New("interrupted")
+)
+
+// EventType is an event type during the runner's run call.
+type EventType int
+
+// String returns a string representation of the EventType.
+func (e EventType) String() string {
+	switch e {
+	case EventTypeStarted:
+		return "started"
+	case EventTypeCmdStarted:
+		return "cmd_started"
+	case EventTypeCmdFinished:
+		return "cmd_finished"
+	case EventTypeFinished:
+		return "finished"
+	default:
+		return strconv.Itoa(int(e))
+	}
+}
+
+// ParseEventType parses an EventType from it's string representation.
+func ParseEventType(eventTypeString string) (EventType, error) {
+	switch eventTypeString {
+	case "started":
+		return EventTypeStarted, nil
+	case "cmd_started":
+		return EventTypeCmdStarted, nil
+	case "cmd_finished":
+		return EventTypeCmdFinished, nil
+	case "finished":
+		return EventTypeFinished, nil
+	default:
+		return 0, fmt.Errorf("unknown EventType string: %s", eventTypeString)
+	}
+}
+
+// Event is an event that happens during the runner's run call.
+type Event interface {
+	Type() EventType
+	Time() time.Time
+	Fields() map[string]string
+}
+
+// EventHandler handles events.
+type EventHandler func(Event)
+
+// Timer returns the current time.
+type Timer func() time.Time
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runnerOptions)
+
+// WithFastFail returns a RunnerOption that will return error fun
+// Run as soon as one of the commands fails.
+func WithFastFail() RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.FastFail = true
+	}
+}
+
+// WithMaxConcurrentCmds returns a RunnerOption that will make the
+// Runner only run maxConcurrentCmds at once, or unlimited if 0.
+func WithMaxConcurrentCmds(maxConcurrentCmds int) RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.MaxConcurrentCmds = maxConcurrentCmds
+	}
+}
+
+// WithEventHandler returns a RunnerOption that will use the
+// given EventHandler.
+func WithEventHandler(eventHandler EventHandler) RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.EventHandler = eventHandler
+	}
+}
+
+// WithTimer returns a RunnerOption that will make the Runner
+// use the given Timer.
+func WithTimer(timer Timer) RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.Timer = timer
+	}
+}
+
+// Runner runs the commands.
+type Runner interface {
+	// Run the commands.
+	//
+	// Return error if there was an initialization error, or any of
+	// the running commands returned with a non-zero exit code.
+	Run(cmds []*exec.Cmd) error
+}
+
+// NewRunner returns a new Runner.
```

Comment:
I think if you were to have a `NewTestRunner` function later, this would be pretty clear - I don't want to special-case the name for the default Runner.